File: MATLAB Files/Inference.m
classdef Inference
    
    properties
        grid_size
        discount
        robot_state
        beta
        thetas
        theta_prior
        %state_prior
        policies
    end
    
    
    methods
        
        function obj = Inference(grid_size, discount, robot_state)
            obj.grid_size = grid_size;
            obj.discount = discount;
            obj.robot_state = robot_state;
            
            obj.beta = 10;
            obj.thetas = obj.beta * obj.generateThetas();
            obj.theta_prior = ones(1, length(obj.thetas))/length(obj.thetas);  % Uniform prior
%             obj.state_prior = ones(1, obj.grid_size * obj.grid_size)/(obj.grid_size^2);
%             obj.state_prior = reshape(obj.state_prior, [obj.grid_size, obj.grid_size]);
            
            obj.policies = ["north", "south", "east", "west", "northeast", ...
                "northwest", "southeast", "southwest", "exit"];
        end
        
        
        function list_binary_arrays = generateThetas(obj)
            n = obj.grid_size^2;
            max_number_represented = 2^n - 1;
            numbers = 0:max_number_represented;
            list_binary_strings = dec2bin(numbers);
            list_binary_arrays = zeros(obj.grid_size, obj.grid_size, length(numbers));

            for i = 1:length(numbers)
                character_array = reshape(list_binary_strings(i, :), [obj.grid_size, obj.grid_size]);

                for j = 1:obj.grid_size
                    for k = 1:obj.grid_size
                        character_array(j, k) = str2double(character_array(j, k));
                    end
                end

                list_binary_arrays(:, :, i) = character_array;
            end

        end
        
        
        function probability = humanModel(obj, policy_index, theta)
            [value, q_value, optimal_policies] = valueIteration(theta, obj.discount);
            exp_q_vals = zeros(1, length(obj.policies));
            for i = 1:length(obj.policies)
                exp_q_vals(i) = exp(-q_value(obj.robot_state(1), obj.robot_state(2), i));
            end

            sum = 0;
            for i = 1:length(exp_q_vals)
                if ~isnan(exp_q_vals(i))
                    sum = sum + exp_q_vals(i);
                end
            end

            exp_q_vals = exp_q_vals/sum;
            probability = exp_q_vals(policy_index);

        end
        
        
        function probability = stateGivenTheta(obj, theta, state_row, state_col)
            if theta(state_row, state_col) == 0
                probability = 0;
            else
                nonzero_ind = find(theta);
                cardinality = length(nonzero_ind);
                probability = 1/cardinality;
            end
        end
        
        
        function dstb = stateInference(obj, policy_index)
            tic
            dstb = zeros(obj.grid_size, obj.grid_size);
            for i = 1:obj.grid_size
                for j = 1:obj.grid_size
                    for k = 1:length(obj.thetas)
                        theta = obj.thetas(:, :, k);
                        state_row = i;
                        state_col = j;
                        numerator = obj.stateGivenTheta(theta, state_row, state_col) * ...
                            obj.humanModel(policy_index, theta) * obj.theta_prior(k);
                        if ~isnan(numerator)
                            dstb(state_row, state_col) = dstb(state_row, state_col) + numerator;
                        end
                    end
                end
            end
            
            sum_dstb = sum(sum(dstb));
            dstb = dstb/sum_dstb;
            
            toc
            
        end
        
        
        function dstb = sampling(obj, policy_index, num_samples)
            % Likelihood Weighting
            tic
            dstb = zeros(obj.grid_size, obj.grid_size);
            samples_list = zeros(num_samples, 3);
            weights = zeros(1, num_samples);
            for i = 1:num_samples
                theta_index = randi([1, length(obj.thetas)]); %sample from prior(theta) (uniform)
                theta = obj.thetas(:, :, theta_index);
                nonzero_ind = find(theta);
                [row, col] = ind2sub([obj.grid_size, obj.grid_size], nonzero_ind);
                if ~isempty(row)
                    samples_list(i, 1) = theta_index;
                    rand_index = randi([1, length(row)]);
                    samples_list(i, 2) = row(rand_index); %sample from stateGivenTheta (uniform)
                    samples_list(i, 3) = col(rand_index); %sample from stateGivenTheta (uniform)
                    weights(i) = obj.humanModel(policy_index, theta);
                end
                 
            end
            
            for i = 1:obj.grid_size
                for j = 1:obj.grid_size
                    for k = 1:num_samples
                        if samples_list(k, 2) == i && samples_list(k, 3) == j
                            if ~isnan(weights(k))
                                dstb(i, j) = dstb(i, j) + weights(k);
                            end
                        end
                    end
                end
            end
            
            sum_dstb = sum(sum(dstb));
            dstb = dstb/sum_dstb;
            
            toc
            
        end
        
        
%         function probability = state_inference1(obj, policy_index, state_row, state_col)
%             probability = 0;
%             for i = 1:length(obj.thetas)
%                 theta = obj.thetas(:, :, i);
%                 prob = humanModel(obj, policy_index, theta);
%                 if ~isnan(prob)
%                     prob = prob * theta(state_row, state_col)/obj.beta * ...
%                         obj.state_prior(state_row, state_col);
%                     probability = probability + prob;
%                 end
%             end
%         end
%         
%         
%         function grid = visualize_state_dstb(obj, policy_index)
%             grid = zeros(obj.grid_size);
%             for i = 1:obj.grid_size
%                 for j = 1:obj.grid_size
%                     grid(i, j) = obj.state_inference1(policy_index, i, j);
%                 end
%             end
%         end
 

        function probability = theta_inference(obj, policy_index, theta_index)
            prob_all_thetas = zeros(1, length(obj.thetas));
            for i = 1:length(prob_all_thetas)
                prob_all_thetas(i) = obj.humanModel(policy_index, obj.thetas(:, :, i)) * obj.theta_prior(i);
            end
            
            sum = 0;
            for i = 1:length(prob_all_thetas)
                if ~isnan(prob_all_thetas(i))
                    sum = sum + prob_all_thetas(i);
                end
            end
         
            prob_all_thetas = prob_all_thetas/sum;
            probability = prob_all_thetas(theta_index);
        end
        
        
        function dstb_modified = thetaInferenceDistribution(obj, policy_index)
            dstb = zeros(1, length(obj.thetas));
            for i = 1:length(obj.thetas)
                dstb(i) = obj.theta_inference(policy_index, i);
            end
            
            dstb_modified = zeros(1, length(dstb));
            for i = 1:length(dstb)
                if ~isnan(dstb(i))
                    dstb_modified(i) = dstb(i);
                end
            end
            
            plot(1:length(dstb), dstb_modified);
            xlabel('Thetas');
            ylabel('Probability');
            title('P(theta | u_H, x_R)');
            
        end
        
        
%         function grid = visualize(obj, dstb)
%             grid = zeros(obj.grid_size);
%             for i = 1:length(dstb)
%                 theta = obj.thetas(:, :, i);
%                 nonzero_ind = find(theta);
%                 [row, col] = ind2sub([obj.grid_size, obj.grid_size], nonzero_ind);
%                 for j = 1:length(row)
%                     grid(row(j), col(j)) = grid(row(j), col(j)) + dstb(i);
%                 end
%             end
%         end
        
        
        
    end
end

File: Inference.py
import matplotlib.pyplot as plt
import numpy as np
import random
import time
from value_iteration import ValueIteration
from value_iteration_human_robot import ValueIterationHumanRobot

class Inference:

    def __init__(self, grid_size, discount, robot_state, beta, robot_goal, robot_action, obs_sizes):
        """
        :param grid_size: [num_rows, num_cols]
        :param discount: Discount factor for value iteration
        :param robot_state: Where is the robot currently
        :param beta: Human model "rationality" parameter
        :param robot_goal: Where is the robot going, by the human's knowledge
        :param robot_action: What is the robot's next action, as the human predicts
        :param obs_sizes: Dimensions of obstacles to be included in the inference
        """
        self.grid_size = grid_size
        self.discount = discount
        self.robot_state = robot_state
        self.beta = beta

        self.robot_goal = robot_goal
        if len(self.robot_goal) == 0:
            self.no_goal = True
        else:
            self.no_goal = False

        self.robot_policies = ["north", "south", "east", "west", "northeast", "northwest", "southeast", "southwest", "exit"]
        self.human_policies = ["north", "south", "east", "west", "northeast", "northwest", "southeast", "southwest", "none", "exit"]

        self.robot_action = robot_action

        self.thetas = []

        self.obs_sizes = obs_sizes
        for obs_size in obs_sizes:
            self.thetas.extend(self.generate_parametrized_thetas(size_obstacle=obs_size))

        self.prior = np.ones(len(self.thetas))/len(self.thetas)



    def generate_all_thetas(self):
        n = self.grid_size[0] * self.grid_size[1]
        max_number_represented = 2 ** n - 1
        numbers = range(max_number_represented + 1)
        list_binary_strings = [np.binary_repr(number, width=self.grid_size[0] * self.grid_size[1]) for number in numbers]
        list_binary_arrays = []

        for i in range(len(numbers)):
            list_binary_strings[i] = np.array([int(elem) for elem in list_binary_strings[i]])
            list_binary_arrays.append(np.reshape(list_binary_strings[i], [self.grid_size[0], self.grid_size[1]]))

        return np.array(list_binary_arrays)


    def generate_parametrized_thetas(self, size_obstacle):
        if max(size_obstacle) >= min(self.grid_size[0], self.grid_size[1]):
            print("Obstacle cannot be larger than grid!")
            return
        obs = np.ones((size_obstacle[0], size_obstacle[1]))
        zeros = np.zeros((size_obstacle[0], self.grid_size[1] - size_obstacle[1]))
        block_row = np.hstack([obs, zeros])
        rolls = [block_row]
        for i in range((self.grid_size[1] + 1 - size_obstacle[1]) - 1):
            block_row = np.roll(block_row, 1)
            rolls.append(block_row)

        zeros = np.zeros((self.grid_size[0] - size_obstacle[0], self.grid_size[1]))
        preliminary_thetas = [np.vstack([roll, zeros]) for roll in rolls]
        thetas = []
        for i in range(len(preliminary_thetas)):
            theta = preliminary_thetas[i]
            if self.no_goal:
                thetas.append(np.copy(theta))
            else:
                if theta[self.robot_goal[0]][self.robot_goal[1]] == 0:
                    theta = -theta
                    theta[self.robot_goal[0]][self.robot_goal[1]] = 1
                    thetas.append(np.copy(theta))
                    theta[self.robot_goal[0]][self.robot_goal[1]] = 0
                    theta = -theta
            for j in range((self.grid_size[0] + 1 - size_obstacle[0]) - 1):
                theta = np.roll(theta, 1, axis=0)
                if self.no_goal:
                    thetas.append(np.copy(theta))
                else:
                    if theta[self.robot_goal[0]][self.robot_goal[1]] == 0:
                        theta = -theta
                        theta[self.robot_goal[0]][self.robot_goal[1]] = 1
                        thetas.append(np.copy(theta))
                        theta[self.robot_goal[0]][self.robot_goal[1]] = 0
                        theta = -theta

        t = np.zeros((self.grid_size[0], self.grid_size[1]))
        t[1, 1] = -1
        t[2, 2] = -1
        t[self.robot_goal[0], self.robot_goal[1]] = 1
        thetas.append(t)

        t = np.zeros((self.grid_size[0], self.grid_size[1]))
        t[1, 1] = -1
        t[2, 2] = -1
        t[1, 2] = -1
        t[2, 1] = -1
        t[self.robot_goal[0], self.robot_goal[1]] = 1
        thetas.append(t)

        return thetas


    def human_model(self, policy_index, theta, final_value_param=10):
        """
        :param policy_index: Index of policy the human inputted according to self.policies
        :param theta: Potential occupancy grid
        :param final_value_param: Reward upon reaching goal state
        :return: Probability human gives the policy given theta, which is a softmax on
        the optimal policy given by value iteration
        """
        if self.no_goal:
            if self.robot_action == -1:
                return self.human_model_no_goal(policy_index, theta, final_value_param)
            else:
                return self.human_model_nogoal_uR(policy_index, theta, final_value_param)
        else:
            if self.robot_action == -1:
                return self.human_model_goal(policy_index, theta, final_value_param)
            else:
                return self.human_model_goal_uR(policy_index, theta, final_value_param)
                #return self.human_model_goal_uR_alternate(policy_index, theta, final_value_param)


    def human_model_no_goal(self, policy_index, theta, final_value_param):
        valiter = ValueIteration(self.grid_size)
        final_value = theta * final_value_param
        value, q_value, optimal_policies = valiter.value_iteration(final_value, self.discount)
        exp_q_vals = np.zeros(len(valiter.policies))
        for i in range(len(valiter.policies)):
            exp_q_vals[i] = np.exp(-self.beta * q_value[self.robot_state[0], self.robot_state[1], i])

        sum_exp = 0
        for i in range(len(exp_q_vals)):
            if not np.isnan(exp_q_vals[i]):
                sum_exp += exp_q_vals[i]

        exp_q_vals /= sum_exp

        return exp_q_vals[policy_index]


    def human_model_goal(self, policy_index, theta, final_value_param):
        valiter = ValueIteration(self.grid_size)
        final_value = theta * final_value_param
        value, q_value, optimal_policies = valiter.value_iteration(final_value, self.discount)
        exp_q_vals = np.zeros(len(valiter.policies))
        for i in range(len(valiter.policies)):
            exp_q_vals[i] = np.exp(self.beta * q_value[self.robot_state[0], self.robot_state[1], i])

        sum_exp = 0
        for i in range(len(exp_q_vals)):
            if not np.isnan(exp_q_vals[i]):
                sum_exp += exp_q_vals[i]

        exp_q_vals /= sum_exp

        return exp_q_vals[policy_index]


    def human_model_nogoal_uR(self, policy_index, theta, final_value_param):
        valiter = ValueIterationHumanRobot(self.grid_size)
        final_value = theta * final_value_param
        value, q_value, optimal_policies_human, optimal_policies_robot = valiter.value_iteration(final_value, self.discount)
        exp_q_vals = np.zeros(len(valiter.human_policies))
        for i in range(len(valiter.human_policies)):
            exp_q_vals[i] = np.exp(-self.beta * q_value[self.robot_state[0], self.robot_state[1], i, self.robot_action])

        sum_exp = 0
        for i in range(len(exp_q_vals)):
            if not np.isnan(exp_q_vals[i]):
                sum_exp += exp_q_vals[i]

        exp_q_vals /= sum_exp

        return exp_q_vals[policy_index]


    def human_model_goal_uR(self, policy_index, theta, final_value_param):
        valiter = ValueIterationHumanRobot(self.grid_size)
        final_value = theta * final_value_param
        value, q_value, optimal_policies_human, optimal_policies_robot = valiter.value_iteration(final_value, self.discount)
        exp_q_vals = np.zeros(len(valiter.human_policies))
        for i in range(len(valiter.human_policies)):
            exp_q_vals[i] = np.exp(self.beta * q_value[self.robot_state[0], self.robot_state[1], i, self.robot_action])

        sum_exp = 0
        for i in range(len(exp_q_vals)):
            if not np.isnan(exp_q_vals[i]):
                sum_exp += exp_q_vals[i]

        exp_q_vals /= sum_exp

        return exp_q_vals[policy_index]


    def human_model_goal_uR_alternate(self, policy_index, theta, final_value_param):
        valiter = ValueIteration(self.grid_size)
        final_value = theta * final_value_param
        value, q_value, optimal_policies = valiter.value_iteration(final_value, self.discount)
        exp_q_vals = np.zeros(len(valiter.policies))
        for i in range(len(valiter.policies)):
            diff_q = q_value[self.robot_state[0], self.robot_state[1], i] - \
                     q_value[self.robot_state[0], self.robot_state[1], self.robot_action]
            exp_q_vals[i] = np.exp(self.beta * diff_q)

        sum_exp = 0
        for i in range(len(exp_q_vals)):
            if not np.isnan(exp_q_vals[i]):
                sum_exp += exp_q_vals[i]

        exp_q_vals /= sum_exp

        return exp_q_vals[policy_index]


    def exact_inference(self, policy_index, visualize=True):
        t0 = time.time()

        dstb = np.zeros(len(self.thetas))
        for i in range(len(self.thetas)):
            theta = self.thetas[i]
            theta = theta.astype('float')
            human_model_prob = self.human_model(policy_index, theta)
            dstb[i] = human_model_prob * self.prior[i]

        sum_dstb = 0
        for i in range(len(dstb)):
            if not np.isnan(dstb[i]):
                sum_dstb += dstb[i]
            else:
                dstb[i] = 0

        dstb /= sum_dstb

        t1 = time.time()

        if visualize:
            self.visualizations(dstb)

        self.prior = dstb

        return dstb, t1-t0


    def visualizations(self, dstb):
        dstb_states = np.zeros((self.grid_size[0], self.grid_size[1]))
        for i in range(self.grid_size[0]):
            for j in range(self.grid_size[1]):
                max_prob = 0
                for k in range(len(self.thetas)):
                    theta = self.thetas[k]
                    if self.no_goal:
                        val = 1
                    else:
                        val = -1
                    if theta[i][j] == val:
                        #max_prob += dstb[k]
                        if dstb[k] > max_prob:
                            max_prob = dstb[k]
                dstb_states[i][j] = max_prob

        # Belief bar chart
        plt.figure()
        plt.bar(range(len(self.thetas)), dstb)
        plt.title("Belief over Theta")
        plt.ylabel("P(theta|u_H)")

        # Trying to visualize probabilities per grid cell
        plt.figure()
        plt.imshow(dstb_states, cmap='hot')
        plt.colorbar()
        plt.scatter(self.robot_state[1], self.robot_state[0], s=50, c='b')

        # Subplots showing probabilities of individual thetas
        sorted_ind = np.argsort(dstb)
        counter = 0
        while counter < len(dstb):
            fig, axs = plt.subplots(3, 3)
            for i in range(3):
                if counter >= len(dstb):
                    break
                for j in range(3):
                    if counter >= len(dstb):
                        break
                    if self.no_goal:
                        axs[i, j].imshow(self.thetas[sorted_ind[counter]], cmap='hot')
                    else:
                        ind = np.where(self.thetas[sorted_ind[counter]] == 1)
                        theta_show = np.copy(self.thetas[sorted_ind[counter]])
                        theta_show[ind[0], ind[1]] = 0
                        theta_show = -theta_show
                        axs[i, j].imshow(theta_show, cmap='hot')

                    axs[i, j].scatter(self.robot_state[1], self.robot_state[0], s=50, c='b')
                    axs[i, j].set_title(dstb[sorted_ind[counter]], fontsize = 8)

                    counter += 1

        plt.show()


        # for i in range(len(dstb)):
        #     if self.no_goal:
        #         plt.imshow(self.thetas[i], cmap='hot')
        #     else:
        #         ind = np.where(self.thetas[i] == 1)
        #         theta_show = np.copy(self.thetas[i])
        #         theta_show[ind[0], ind[1]] = 0
        #         theta_show = -theta_show
        #         plt.imshow(theta_show, cmap='hot')
        #
        #     plt.scatter(self.robot_state[1], self.robot_state[0], s=50, c='b')
        #     plt.title(dstb[i])
        #     plt.show()


    def update_thetas(self, goal):
        self.robot_goal = goal
        self.thetas = []
        for obs_size in self.obs_sizes:
            self.thetas.extend(self.generate_parametrized_thetas(size_obstacle=obs_size))







